Only delete known Maester tests with Update-MaesterTests

SamErde opened the pull request:
Checks the known Maester test names from the installed module path and only deletes those child object names instead of everything in $Path. Provides a resolution for #173.

merill commented:
This is good Sam. Thanks. Do you have thoughts on how to handle scenarios where we delete/retire tests?

SamErde commented:
Yes, I do have some early thoughts on how to handle removed or retired tests. I was also thinking of adding a test update process that can be separate from the module update process (an -Online parameter for updating tests).

Handling removed tests could be handled in a few ways:
- Maintain an inventory of active tests, either manually as they change or via a script that checks test files/folders during a build step.
- Leave the test files, but append ".Retired" (or similar) to the .PS1 and .MD file extension.
- Add status metadata in each of the test files themselves, with logic in the Invoke and Update functions to skip/remove retired tests.
- Move them to a "deprecated" folder and exclude that folder when invoking the tests.
- Probably other possibilities as well.

I believe any of these ideas could benefit from an optional online update method that doesn't depend on the module being updated, but I'm sure there are factors that need to be considered.

When considering module download size, there will be benefits to simply removing retired tests from new releases.

merill commented:
I'm thinking we can add a hashtable to update-maester with a list of all .tests.ps1 files that we've removed. Then it can do a targeted delete of just those.

SamErde commented:
Would you generate the hash table in some automatic fashion by comparing builds or releases, or just add the information by hand when necessary?

SamErde commented:
> I'm thinking we can add a hashtable to update-maester with a list of all .tests.ps1 files that we've removed. Then it can do a targeted delete of just those.

A thought: using a hash table to selectively remove retired tests will be a change to the current wipe/replace approach. If it works, would you like to test and merge this PR, and then follow-up with a separate PR to rethink how updates/changes to tests can be handled? This PR includes:

- Error handling for directory creation and file copy operations.
- Relative path expansion and validation.
- Enhanced verbose output during the new/copy operations.
- Checks to ensure that the only things that get deleted will be files and folders that match the names present in the module's installed `maester-tests` directory.
- Replaced `exit` with `return` statements.
- Simplification of the function and removal of a redundant line of code.
- Added comment-based help.

merill reviewed (approved):
💯% agree. This fixes a lot of the issues and makes the process non-destructive.

Also, I doubt we are going to delete any test folders soon so that can come later when we do get to that point.

SamErde commented:
OK, thanks!
